### PR TITLE
feat/Datepicker switches to native table

### PR DESCRIPTION
**Pull request by rymeskar**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ yarn change`

#### Description of changes
* Datepicker switches to native table
* Adding CalendarGrid component slot

#### Focus areas to test

(optional)

### CODE REVIEW

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx [hunk 1]
```diff
@@ -324,43 +331,39 @@ export const DatepickerCalendar: ComponentWithAs<'div', DatepickerCalendarProps>
...
+          },
+        }),
+      })}
+      {createShorthand(DatepickerCalendarGrid, calendarGrid, {
+        defaultProps: () =>
+          getA11yProps('calendarGrid', {
+            content: (
+              <>
+                <thead>
+                  <tr key={0}>
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:366]: Please let's use `DatepickerCalendarGridRow` here as well

File: packages/fluentui/accessibility/src/behaviors/Datepicker/datepickerCalendarCellBehavior.ts [hunk 2]
```diff
@@ -5,14 +5,19 @@ import { Accessibility } from '../../types';
  * Behavior for a datepicker calendar grid.
```
Comment (pompomon) [inline packages/fluentui/accessibility/src/behaviors/Datepicker/datepickerCalendarCellBehavior.ts:5]: Please could you add justification why we are adding role "none" and "gridcell" to the respective slots?

File: packages/fluentui/accessibility/src/behaviors/Datepicker/datepickerCalendarCellBehavior.ts [hunk 3]
```diff
@@ -5,14 +5,19 @@ import { Accessibility } from '../../types';
...
  * @specification
  * Adds attribute 'aria-selected' based on the property 'selected' to 'root' slot.
+ * Adds role='gridcell'.
```
Comment (pompomon) [inline packages/fluentui/accessibility/src/behaviors/Datepicker/datepickerCalendarCellBehavior.ts:8]: Please could you add a specification for role on tableCell? And I think the comment for `gridcell` role also needs to have the slot specified.

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx [hunk 4]
```diff
@@ -371,14 +369,24 @@ export const DatepickerCalendar: ComponentWithAs<'div', DatepickerCalendarProps>
...
-    </Ref>
+                  </tr>
+                </thead>
+                <tbody>
+                  {_.map(visibleGrid, (week, idx) =>
+                    createShorthand(DatepickerCalendarGridRow, calendarGridRow, {
+                      defaultProps: () =>
+                        getA11yProps('calendarGridRow', {
+                          children: renderWeekRow(week),
+                          key: idx + 1,
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:380]: Do we need the the hardcoded `key` here?

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerVariables.ts [hunk 5]
```diff
@@ -29,9 +34,11 @@ export interface DatepickerVariables {
 }
 
 export const datepickerVariables = (siteVars): DatepickerVariables => ({
+  calendarGridWidth: pxToRem(DAYS_IN_WEEK * 30),
+
   calendarCellBorder: 'none',
-  calendarCellHeight: pxToRem(32),
-  calendarCellWidth: pxToRem(32),
+  calendarCellHeight: pxToRem(30),
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerVariables.ts:40]: Please why did you change the height from 32px to 30px?

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerVariables.ts [hunk 6]
```diff
@@ -29,9 +34,11 @@ export interface DatepickerVariables {
...
+  calendarCellWidth: '100%',
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerVariables.ts:41]: And why calendarCell width is always 100% and headerCell width below is calculated?

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendarCell.tsx [hunk 7]
```diff
@@ -124,17 +124,19 @@ export const DatepickerCalendarCell = compose<
...
-          className: classes.root,
-          onClick: handleClick,
-          ref,
-          disabled,
-          ...unhandledProps,
-        })}
-      >
-        {content}
-      </ElementType>
+      <td {...getA11yProps('tableCell', {})}>
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendarCell.tsx:127]: I'm not 100% sure on that one - please @layershifter could you take a look at that?

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx (hunk 4 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:380]: Yeah, otherwise framework complains:
image

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerVariables.ts (hunk 5 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerVariables.ts]: Fixed layout :)

File: packages/date-time-utilities/src/dateGrid/getDayGrid.ts [hunk 8]
```diff
@@ -54,14 +54,18 @@ export const getDayGrid = (options: IDayGridOptions): IDay[][] => {
...
-    selectedDate,
-    selectedDateRangeType,
-    firstDayOfWeek,
-    workWeekDays,
-    daysToSelectInDayView,
-  );
-  selectedDates = getBoundedDateRange(selectedDates, minDate, maxDate);
+  let selectedDates: Date[] = [];
+
+  if (!!selectedDate) {
```
Comment (pompomon) [inline packages/date-time-utilities/src/dateGrid/getDayGrid.ts:59]: Please why do we need that check?

Comment (rymeskar) [reply packages/date-time-utilities/src/dateGrid/getDayGrid.ts:59]: If we don't do this check, then the dayGrid will always return at least one selected day. That should not be the case initially though.

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendarCell.tsx (hunk 7 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendarCell.tsx]: Switched to slots. :)

File: packages/date-time-utilities/src/dateGrid/getDayGrid.ts (hunk 8 above)
Comment (pompomon) [reply packages/date-time-utilities/src/dateGrid/getDayGrid.ts:59]: @lorejoh12 , please would that break anything on v7 Datepicker?

Comment (lorejoh12) [reply packages/date-time-utilities/src/dateGrid/getDayGrid.ts:59]: I think this should be fine. Normal usage that doesn't pass in an initial selected date wouldn't necessarily be expected to have a selected date range. Also from testing the examples in the deployed test page, the existing examples that do have an initially selected date (week selection, month selection) still initially render fine, so I think this should be okay.

Signing off for this specific change

**Review by lorejoh12 — APPROVED** [review]

File: packages/date-time-utilities/src/dateGrid/getDayGrid.ts [hunk 9]
```diff
@@ -54,14 +54,18 @@ export const getDayGrid = (options: IDayGridOptions): IDay[][] => {
...
+    selectedDates = getDateRangeArray(
+      selectedDate,
+      selectedDateRangeType,
+      firstDayOfWeek,
+      workWeekDays,
+      daysToSelectInDayView,
+    );
+    selectedDates = getBoundedDateRange(selectedDates, minDate, maxDate);
+  }
```
Comment (layershifter) [inline packages/date-time-utilities/src/dateGrid/getDayGrid.ts:68]: I suggest to add a unit test for this case

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 10]
```diff
@@ -322,55 +325,51 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+      {...getA11yProps('root', {
+        className: classes.root,
+        ...unhandledProps,
+      })}
+    >
+      {!props.buttonOnly &&
+        createShorthand(Input, input, {
+          defaultProps: () =>
+            getA11yProps('input', {
+              placeholder: props.inputPlaceholder,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:339]:
```suggestion
              placeholder: props.inputPlaceholder,
```

Should be this removed? It can be defined via shorthand `<Datepicker input={{ placeholder }} />`

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 11]
```diff
@@ -322,55 +325,51 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+              disabled: props.disabled,
+              error: !!error,
+              value: formattedDate,
+              readOnly: !allowManualInput,
+              required: props.required,
+              ref: inputRef,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:345]: This should be in `overrideProps` and perform merge as if I will pass: `<Datepicker input={{ ref }} />` it will be broken

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 12]
```diff
@@ -322,55 +325,51 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
+        })}
+      {createShorthand(Popup, popup, {
+        defaultProps: () => ({
+          open: openState && !props.disabled,
+          content: calendarElement,
+          trapFocus: {
+            disableFirstFocus: true,
+          },
+          trigger: triggerButtonElement,
+          target: !props.buttonOnly ? inputRef.current : null,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:358]:
```suggestion
          target: props.buttonOnly ? null : inputRef.current,
```

nit

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx [hunk 13]
```diff
@@ -322,55 +325,51 @@ export const Datepicker: ComponentWithAs<'div', DatepickerProps> &
...
-          })}
-        </ElementType>,
-      )}
-    </Ref>
+          overrideProps: overrideInputProps,
+        })}
+      {createShorthand(Popup, popup, {
+        defaultProps: () => ({
+          open: openState && !props.disabled,
+          content: calendarElement,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:353]: I expect that `trigger`, `target` & `content` will be in `overrideProps`, what should happen if I will specify `<Datepicker popup={{ content: 'foo' }} />`?

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx [hunk 14]
```diff
@@ -295,90 +302,107 @@ export const DatepickerCalendar: ComponentWithAs<'div', DatepickerCalendarProps>
...
+      defaultProps: () =>
+        getA11yProps('calendarCell', {
+          content: day.date,
+          'aria-label': formatMonthDayYear(day.originalDate, dateFormatting),
+          selected: day.isSelected,
+          disabled: !day.isInBounds,
+          quiet: !day.isInMonth,
+          today: compareDates(day.originalDate, props.today ?? new Date()),
+          ref: compareDates(gridNavigatedDate, day.originalDate) ? focusDateRef : null,
+          onFocus: () => setGridNavigatedDate(day.originalDate),
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:329]:
```suggestion
```

This should go to `overrideProps`

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx [hunk 15]
```diff
@@ -295,90 +302,107 @@ export const DatepickerCalendar: ComponentWithAs<'div', DatepickerCalendarProps>
...
+                          createShorthand(DatepickerCalendarHeaderCell, calendarHeaderCell, {
+                            defaultProps: () =>
+                              getA11yProps('calendarHeaderCell', {
+                                'aria-label': days[(dayNumber + firstDayOfWeek) % DAYS_IN_WEEK],
+                                content: shortDays[(dayNumber + firstDayOfWeek) % DAYS_IN_WEEK],
+                                key: dayNumber,
+                              }),
                           }),
+                        ),
+                        key: 0,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:386]:
```suggestion
```

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts [hunk 16]
```diff
@@ -1,63 +1,42 @@
...
+  ...(p.quiet && {
+    color: v.calendarCellQuietColor,
+  }),
+
+  ...(p.selected && {
+    color: v.calendarCellSelectedColor,
+    backgroundColor: v.calendarCellSelectedBackgroundColor,
+  }),
+
+  ...(!p.disabled && {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts:15]: Let's avoid negative conditions in styles

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts [hunk 17]
```diff
@@ -1,63 +1,42 @@
 import { ComponentSlotStylesPrepared, ICSSInJSStyle } from '@fluentui/styles';
 import { DatepickerVariables } from './datepickerVariables';
 import { DatepickerCalendarCellStylesProps } from '../../../../components/Datepicker/DatepickerCalendarCell';
-import { getBorderFocusStyles } from '../../getBorderFocusStyles';
+
+const cellStyles = (p: DatepickerCalendarCellStylesProps, v: DatepickerVariables) => ({
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts:5]: What is a purpose of having `cellStyles` function? I see only a single usage of it

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarHeaderCellStyles.ts [hunk 18]
```diff
@@ -9,7 +9,19 @@ export const datepickerCalendarHeaderCellStyles: ComponentSlotStylesPrepared<
...
+
       fontWeight: v.calendarHeaderCellFontWeight,
+
+      height: v.calendarHeaderCellHeight,
+      width: v.calendarHeaderCellWidth,
+
+      paddingTop: v.calendarHeaderCellPaddingTop,
+      paddingBottom: v.calendarHeaderCellPaddingBottom,
+      paddingLeft: v.calendarHeaderCellPaddingLeft,
+      paddingRight: v.calendarHeaderCellPaddingRight,
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarHeaderCellStyles.ts:22]:
```suggestion
      padding: v.calendarHeaderCellPadding,
```

Let's have this as a single padding

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellButtonStyles.ts [hunk 19]
```diff
@@ -0,0 +1,76 @@
+import { ComponentSlotStylesPrepared, ICSSInJSStyle } from '@fluentui/styles';
+import { DatepickerVariables } from './datepickerVariables';
+import { DatepickerCalendarCellButtonStylesProps } from '../../../../components/Datepicker/DatepickerCalendarCellButton';
+import { getBorderFocusStyles } from '../../getBorderFocusStyles';
+
+const cellStyles = (p: DatepickerCalendarCellButtonStylesProps, v: DatepickerVariables) => ({
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellButtonStyles.ts:6]: The same is there, `cellStyles` are not reused

File: packages/date-time-utilities/src/dateGrid/getDayGrid.ts (hunk 9 above)
Comment (rymeskar) [reply packages/date-time-utilities/src/dateGrid/getDayGrid.ts:68]: Added

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 10 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:339]: There is no specific reason for keeping the placeholder explicit. Will switch.

FYI: @pompomon

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 11 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:345]: Curious: 
In what scenarios might the user want to override the inputRef? 

Except for the overrideProps section within Input, how would we get access to the user's entered value (context: Popup.onOpenChange; Popup.target)?

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 13 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:353]: Good point. Will also change position and align. 

I was though wondering. Does it make sense to override the 'content'? This is the actual calendar that can already be overriden through the calendar slot.

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx (hunk 14 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:329]: Does it make sense to override props that are intended to be used per day (both ref and onFocus); though the user does not receive the 'day' information in the exposed interface?

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts (hunk 16 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts:15]: I see.

What is then the preferred coding style for creating styling properties in case a given boolean is true and in case not?

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx [hunk 20]
```diff
@@ -295,90 +302,107 @@ export const DatepickerCalendar: ComponentWithAs<'div', DatepickerCalendarProps>
...
-    </Ref>
+                  })}
+                </thead>
+                <tbody>
+                  {_.map(visibleGrid, (week, idx) =>
+                    createShorthand(DatepickerCalendarGridRow, calendarGridRow, {
+                      defaultProps: () =>
+                        getA11yProps('calendarGridRow', {
+                          children: renderWeekRow(week),
+                          key: idx + 1,
```
[the remark this answers is not in the document]
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:395]: Upon changing to week, I get the following error in runtime.

image

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 11 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:345]: > In what scenarios might the user want to override the inputRef?

User does not want to override a ref, but wants to get a `ref` and then it will cause weird bugs:
https://github.com/microsoft/fluentui/issues/13661
https://github.com/microsoft/fluentui/pull/13711

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 13 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:353]: > Good point. Will also change position and align.

I think that `position` & `align` should stay in default to be overridable to users.

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts (hunk 16 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts:15]: Have always a default value and then override it with a positive condition.

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:395]: Key should be a string, `week` implements `IDay` as I see and it has `.key` prop, will `week.key` work?

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx (hunk 14 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:329]: For `onFocus` you have to `predefinedProps.onFocus`, for `ref` it's the same case as linked in a previous comment

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 11 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:345]: Updated.

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 13 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:353]: Updated

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 10 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:339]: I checked out the implementation again and the placeholder is part of the datepicker localization strings (ICalendarStrings). The user then always has the translations in one spot. 

Does it then from this stance make sense to keep the `inputPlaceholder` prop as is?

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx (hunk 14 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:329]: Updated onFocus. 

Kept the ref. I wanted to ask: should I create a ref prop in the DatepickerCalendarCellButton?

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:395]: updated

File: packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts (hunk 16 above)
Comment (rymeskar) [reply packages/fluentui/react-northstar/src/themes/teams/components/Datepicker/datepickerCalendarCellStyles.ts:15]: Updated.

File: packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx (hunk 10 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/Datepicker.tsx:345]: Let's keep it as it is for now.

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx [hunk 21]
```diff
@@ -326,9 +326,12 @@ export const DatepickerCalendar: ComponentWithAs<'div', DatepickerCalendarProps>
           quiet: !day.isInMonth,
           today: compareDates(day.originalDate, props.today ?? new Date()),
           ref: compareDates(gridNavigatedDate, day.originalDate) ? focusDateRef : null,
-          onFocus: () => setGridNavigatedDate(day.originalDate),
         }),
       overrideProps: (predefinedProps: DatepickerCalendarCellButtonProps): DatepickerCalendarCellButtonProps => ({
+        onFocus: e => {
+          setGridNavigatedDate(day.originalDate);
+          _.invoke(predefinedProps, 'onFocus', e, { ...predefinedProps, value: day });
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx:333]:
```suggestion
          _.invoke(predefinedProps, 'onFocus', e, predefinedProps);
```

File: packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx (hunk 14 above)
Comment (layershifter) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx]: > I create a ref prop in the DatepickerCalendarCellButton

No, `ref` is not a React prop and it should be in props, if you have typings issue, please fix it in exact usage, something like `DatepickerCalendarCellButtonProps & { ref: A_TYPE_THAT_CORRESPONDS }`

Comment (rymeskar) [reply packages/fluentui/react-northstar/src/components/Datepicker/DatepickerCalendar.tsx]: Thanks. Fully updated.

**Review by pompomon — APPROVED** [review]
